Lib/ufo2ft/compilers/baseCompiler.py

```diff
@@ -0,0 +1,306 @@
+                    master_locations = []
+                    for sourceDescriptor in vfDoc.sources:
+                        master_location = sourceDescriptor.name + " at "
+                        master_location += location_to_string(
+                            sourceDescriptor.getFullDesignLocation(vfDoc)
+                        )
+                        master_locations.append(master_location)
+                    master_location_descriptions = "\n".join(master_locations)
+                    raise InvalidDesignSpaceData(
+                        f"No default source; expected default master at {default_location}."
+                        f" Found master locations:\n{master_location_descriptions}"
+                    )
+                vfNameToBaseUfo[vfName] = default_source.font
+                for source in vfDoc.sources:
+                    sourcesToCompile.add(source.name)
+
+        # Match sources to compile to their Descriptor in the original designspace
+        sourcesByName = {}
+        for source in designSpaceDoc.sources:
+            if source.name in sourcesToCompile:
+                sourcesByName[source.name] = source
+
+        originalSources = {}
+
+        # Compile all needed sources in each interpolable subspace to make sure
+        # they're all compatible; that also ensures that sub-vfs within the same
+        # interpolable sub-space are compatible too.
+        for subDoc in interpolableSubDocs:
+            # Only keep the sources that we've identified earlier as need-to-compile
+            subDoc.sources = [s for s in subDoc.sources if s.name in sourcesToCompile]
+            if not subDoc.sources:
+                continue
+
+            # FIXME: Hack until we get a fontTools config module. Disable GPOS
+            # compaction while building masters because the compaction will be undone
+            # anyway by varLib merge and then done again on the VF
+            gpos_compact_value = os.environ.pop(GPOS_COMPACT_MODE_ENV_KEY, None)
+            save_production_names = self.useProductionNames
+            self.useProductionNames = False
+            save_postprocessor = self.postProcessorClass
+            self.postProcessorClass = None
+            try:
+                ttfDesignSpace = self.compile_designspace(subDoc)
+            finally:
+                if gpos_compact_value is not None:
+                    os.environ[GPOS_COMPACT_MODE_ENV_KEY] = gpos_compact_value
+            self.postProcessorClass = save_postprocessor
+            self.useProductionNames = save_production_names
+
+            # Stick TTFs back into original big DS
+            for ttfSource in ttfDesignSpace.sources:
+                sourcesByName[ttfSource.name].font = ttfSource.font
+
+        return vfNameToBaseUfo, originalSources
+
+    def compile_variable(self, designSpaceDoc):
+        # Pop inplace because we'll make a copy at this level so deeper functions
+        # don't need to worry
```

Comment:
the comment is stale and no longer applies since you aren't 'popping' any longer